src/core/stores/VRange.ts

```diff
@@ -0,0 +1,94 @@
+import { VNode, VNodeType } from './VNode';
+
+export enum RangeDirection {
+    BACKWARD = 'BACKWARD',
+    FORWARD = 'FORWARD',
+}
+export enum RelativePosition {
+    BEFORE = 'BEFORE',
+    AFTER = 'AFTER',
+}
+export interface TargetLocation {
+    reference: VNode;
+    relativePosition: RelativePosition;
+}
+export interface VRangeLocation {
+    start: TargetLocation;
+    end: TargetLocation;
+    direction: RangeDirection;
+}
+
+export class VRange {
+    readonly _start = new VNode(VNodeType.RANGE_START);
+    readonly _end = new VNode(VNodeType.RANGE_END);
+    direction: RangeDirection;
+    constructor(direction: RangeDirection = RangeDirection.FORWARD) {
+        this.direction = direction;
+    }
+    /**
+     * Collapse the range. Return self.
+     *
+     * @param [onEnd] true to collapse on the end range
+     */
+    collapse(onEnd = false): VRange {
+        if (onEnd) {
+            return this.setStart(RelativePosition.BEFORE, this._end);
+        } else {
+            return this.setEnd(RelativePosition.AFTER, this._start);
+        }
+    }
+    /**
+     * Move the range to the given location. If no end location is given,
+     * collapse on the start location. A location is given by targetting a
+     * reference VNode and specifying the position in reference to that VNode
+     * ('BEFORE', 'AFTER'), like in an `xpath`. Return self.
+     *
+     * @param start
+     * @param [end]
+     */
+    move(start: TargetLocation, end?: TargetLocation): VRange {
+        this.setStart(start.relativePosition, start.reference);
+        const sameReferences = !end || end.reference.id === start.reference.id;
+        const samePositions = !end || end.relativePosition === start.relativePosition;
+        if (sameReferences && samePositions) {
+            return this.collapse();
+        } else {
+            return this.setEnd(end.relativePosition, end.reference);
+        }
+    }
+    /**
+     * Change the range's direction (forward or backward). Return self.
+     *
+     * @param direction
+     */
+    setDirection(direction: RangeDirection): VRange {
+        this.direction = direction;
```

Comment:
Actually the direction is FORWARD if _start comes before _end, but calculating it would imply traversing the tree every time...

Reply:
Naively (the `isBefore` method should perform an efficient search instead of doing a straight traversal):
```typescript
// VNode.ts

/**
 * Return true if this VNode comes before the given VNode in the pre-order
 * traversal.
 *
 * @param vNode
 */
isBefore(vNode: VNode): boolean {
    const next = this.next();
    if (next && next.id !== vNode.id) {
        return next.isBefore(vNode);
    } else {
        return !!next;
    }
}

// VRange.ts

_direction: RangeDirection;
constructor(direction: RangeDirection = RangeDirection.FORWARD) {
    this._direction = direction;
}
get direction(): RangeDirection {
    if (!this._direction) {
        const isForward = this._start.isBefore(this._end);
        this._direction = isForward ? RangeDirection.FORWARD : RangeDirection.BACKWARD;
    }
    return this._direction;
}
...
setStart(...) {
    ...
    // Invalidate range direction cache
    this._direction = null;
}

setEnd(...) {
    ...
    // Invalidate range direction cache
    this._direction = null;
}
```
Then we can also add `this._direction = RangeDirection.FORWARD;` to `VRange.collapse()` because a collapsed range is always 'forward' (we set the end after the start).
We can also add
```typescript
if (location.direction) {
    this._direction = location.direction;
}
```
to `VRange.set` because when we set a range from the `EventNormalizer` (which is most of the time), we always know what the direction is.

Reply:
(see https://github.com/odoo-dev/jabberwock/commit/e4043b923eb946e42f44936efa1dacba11e6f000)

Reply:
Here's already a bit of a better version of `isBefore` (in particular for when it should returns false):
```typescript
isBefore(vNode: VNode, i = 1): boolean {
    const next = this.next();
    const previousVNode = vNode.previous();
    if (next && next.id !== vNode.id && previousVNode && previousVNode.id !== next.id) {
        return next.isBefore(previousVNode, i + 1);
    } else {
        return !!next && !!previousVNode;
    }
}
At every step it moves the wanted vNode back one step... Combining this with a simultaneous search in the other direction that returns false if it finds it could bring better results.
Also to ponder: can we find a way to do a A* Search on the subset of the tree that is "all the nodes after this"?
```